mithril-common/src/signable_builder/cardano_database.rs

```diff
@@ -0,0 +1,145 @@
+    logging::LoggerExtensions,
+    signable_builder::SignableBuilder,
+    StdResult,
+};
+
+/// This structure is responsible for calculating the message for incremental Cardano database.
+pub struct CardanoDatabaseSignableBuilder {
+    digester: Arc<dyn ImmutableDigester>,
+    logger: Logger,
+    dirpath: PathBuf,
+}
+
+impl CardanoDatabaseSignableBuilder {
+    /// Constructor
+    pub fn new(digester: Arc<dyn ImmutableDigester>, dirpath: &Path, logger: Logger) -> Self {
+        Self {
+            digester,
+            logger: logger.new_with_component_name::<Self>(),
+            dirpath: dirpath.to_owned(),
+        }
+    }
+}
+
+#[async_trait]
+impl SignableBuilder<CardanoDbBeacon> for CardanoDatabaseSignableBuilder {
+    async fn compute_protocol_message(
+        &self,
+        beacon: CardanoDbBeacon,
+    ) -> StdResult<ProtocolMessage> {
+        let merkle_tree = self
+            .digester
+            .compute_merkle_tree(&self.dirpath, &beacon)
+            .await
+            .with_context(|| {
+                format!(
+                    "Cardano Database Signable Builder can not compute merkle tree of '{}'",
+                    &self.dirpath.display()
+                )
+            })?;
+
+        let merkle_root = merkle_tree.compute_root()?.to_hex();
+        info!(
+            self.logger,
+            "Computed Cardano database Merkle root = '{merkle_root}'"
+        );
+
+        let mut protocol_message = ProtocolMessage::new();
+        protocol_message.set_message_part(
+            ProtocolMessagePartKey::CardanoDatabaseMerkleRoot,
+            merkle_root,
+        );
+
+        Ok(protocol_message)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::path::Path;
+
+    use crate::{
+        crypto_helper::{MKTree, MKTreeNode, MKTreeStoreInMemory},
+        digesters::ImmutableDigesterError,
+        entities::{CardanoDbBeacon, ProtocolMessagePartKey},
+        test_utils::TestLogger,
+    };
+
+    use super::*;
+
+    #[derive(Default)]
+    pub struct ImmutableDigesterImpl {
+        digests: Vec<String>,
+    }
+
+    impl ImmutableDigesterImpl {
+        pub fn new(digests: Vec<String>) -> Self {
+            Self { digests }
+        }
+    }
+
+    #[async_trait]
+    impl ImmutableDigester for ImmutableDigesterImpl {
+        async fn compute_digest(
+            &self,
+            _dirpath: &Path,
+            _beacon: &CardanoDbBeacon,
+        ) -> Result<String, ImmutableDigesterError> {
+            Ok("whatever".to_string())
+        }
+
+        async fn compute_merkle_tree(
+            &self,
+            _dirpath: &Path,
+            _beacon: &CardanoDbBeacon,
+        ) -> Result<MKTree<MKTreeStoreInMemory>, ImmutableDigesterError> {
+            let leaves = create_leaves_from_digests(self.digests.clone());
+
+            Ok(MKTree::new(&leaves).unwrap())
+        }
+    }
+
+    fn create_leaves_from_digests(digests: Vec<String>) -> Vec<MKTreeNode> {
```

Comment:
Same remark as above, this function should not be necessary.

Reply:
Done ✅